### PR TITLE
Refactor plugin registration

### PR DESCRIPTION
**Pull request by linus-sherrill**
Refactor plugin registration into a registrar class hierarchy. The registrar collects all common behaviours needed for registering the different groups of plugins. The result is less code in the register source file.

### CODE REVIEW

**Comment by linus-sherrill** [comment]
Jenkins test this please

File: tools/applet_registrar.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef KWIVER_TOOLS_KWIVER_APPLET_REGISTER_H
+#define KWIVER_TOOLS_KWIVER_APPLET_REGISTER_H
+
+#include <vital/plugin_loader/plugin_registrar.h>
+#include <tools/kwiver_applet.h>
```
Comment (mleotta) [inline tools/applet_registrar.h]: I think the framework classes for supporting applets should be moved into `vital` or possibly `vital/applets`, but this could happen in a later branch.  The only tool is the `kwiver` tool.  I don't think the tools directory should be on the include path, but it could contain the .cxx for building the actual kwiver executable.

File: tools/applet_registrar.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+namespace kwiver {
+
+class applet_registrar
```
Comment (mleotta) [inline tools/applet_registrar.h]: Some basic doc strings in this file would be useful.

File: sprokit/src/sprokit/pipeline/process_factory.h [hunk 3]
```diff
@@ -193,6 +194,49 @@ kwiver::vital::plugin_factory_vector_t const& get_process_list();
                                                  typeid( sprokit::process ).name(), \
                                                  sprokit::create_new_process< proc_type > ) )
 
+// ============================================================================
+class process_registrar
```
Comment (mleotta) [inline sprokit/src/sprokit/pipeline/process_factory.h]: Some basic doc strings for this class would be useful.

File: vital/algo/algorithm_factory.h [hunk 4]
```diff
@@ -145,6 +146,38 @@ bool has_algorithm_impl_name(std::string const& type_name,
 #define ADD_ALGORITHM( impl_name, conc_T)                    \
   add_factory( new kwiver::vital::algorithm_factory_0<conc_T>( conc_T::static_type_name(), impl_name ))
 
+// ============================================================================
+class algorithm_registrar
```
Comment (mleotta) [inline vital/algo/algorithm_factory.h]: Some basic doc strings for this class would be useful.

File: vital/plugin_loader/plugin_registrar.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  class plugin_loader;
+} // end namespace vitak
+
+
+/// Class to assist in registering tools.
+class plugin_registrar
+{
+public:
+  /**
+   * @brief Create registrar
```
Comment (mleotta) [inline vital/plugin_loader/plugin_registrar.h]: I don't think we've been very consistent about this, but I think most of KWIVER uses `/// brief description` on the first line instead of `* @brief brief description` in the body and uses `\param` instead of `@param`.  I don't really care too much which style we use, but consistency is more important.

File: vital/plugin_loader/plugin_registrar.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+#if defined PLUGIN_ORG
+    , mod_organization( PLUGIN_ORG )
+#else
+    , mod_organization( "Unspecified" )
+#endif
+    , m_plugin_loader( vpl )
+  {
+  }
+
+  bool is_module_loaded() { return m_plugin_loader.is_module_loaded( mod_name ); }
```
Comment (mleotta) [inline vital/plugin_loader/plugin_registrar.h]: It's probably a good idea to add a `/// brief description` to each of these member functions as well.

File: doc/release-notes/master.txt [hunk 7]
```diff
@@ -16,6 +16,22 @@ Vital
...
+ * Applied a new approach to adding metadata to plugins. The plugin
+   name and description are now specified in the plugin
+   header. Added new plugin registrar class framework to simplify the
+   registration process.
+
+Vital Tools
+
+ * Converted tools to use applet plugin approach.
+
+ * Converting stand-alone tools to reside under the main "kwiver" top
```
Comment (mwoehlke-kitware) [inline doc/release-notes/master.txt]: Not sure this belongs in this PR, be meh... however, shouldn't it be "converted", not "converting"?

File: sprokit/processes/adapters/register_processes.cxx [hunk 8]
```diff
@@ -29,9 +29,9 @@
  */
 
 #include <sprokit/processes/adapters/kwiver_processes_adapter_export.h>
-#include <sprokit/pipeline/process_factory.h>
 
 #include <vital/plugin_loader/plugin_loader.h>
+#include <sprokit/pipeline/process_factory.h>
```
Comment (mwoehlke-kitware) [inline sprokit/processes/adapters/register_processes.cxx]: Eh? Why this change? Shouldn't sprokit headers be kept together?

Actually, shouldn't this be:
```c++
#include "input_adapter_process.h"
#include "output_adapter_process.h"

#include <sprokit/processes/adapters/kwiver_processes_adapter_export.h>
// following newline may or may not be needed/beneficial

#include <sprokit/pipeline/process_factory.h>

#include <vital/plugin_loader/plugin_loader.h>
```
...?

File: sprokit/processes/adapters/register_processes.cxx [hunk 9]
```diff
@@ -46,34 +46,19 @@ extern "C"
 KWIVER_PROCESSES_ADAPTER_EXPORT
 void
 register_factories( kwiver::vital::plugin_loader& vpm )
-
 {
-  static auto const module_name = kwiver::vital::plugin_manager::module_t( "kwiver_processes_adapters" );
+  using namespace sprokit;
```
Comment (mwoehlke-kitware) [inline sprokit/processes/adapters/register_processes.cxx]: This doesn't seem used?

File: sprokit/processes/adapters/register_processes.cxx [hunk 10]
```diff
@@ -46,34 +46,19 @@ extern "C"
...
-
-  fact = vpm.ADD_PROCESS( kwiver::output_adapter_process );
-  fact->add_attribute(  kwiver::vital::plugin_factory::PLUGIN_NAME,  "output_adapter" )
-    .add_attribute(  kwiver::vital::plugin_factory::PLUGIN_MODULE_NAME, module_name )
-    .add_attribute(  kwiver::vital::plugin_factory::PLUGIN_DESCRIPTION,
-                     "Sink process for pipeline.\n\n"
-                     "Accepts data items from pipeline ports. "
-                     "Ports are dynamically created as needed based on connections specified in the pipeline file." )
-    .add_attribute( "no-test", "introspect" ); // do not include in introspection test
+  reg.register_process< kwiver::input_adapter_process > ( process_registrar::no_test );
```
Comment (mwoehlke-kitware) [inline sprokit/processes/adapters/register_processes.cxx]: Pedantic: pretty sure there should be no space between `>(`...

File: sprokit/processes/adapters/register_processes.cxx [hunk 11]
```diff
@@ -46,34 +46,19 @@ extern "C"
...
+  reg.register_process< kwiver::output_adapter_process > ( process_registrar::no_test );
 
   // - - - - - - - - - - - - - - - - - - - - - - -
```
Comment (mwoehlke-kitware) [inline sprokit/processes/adapters/register_processes.cxx]: Now that there isn't a giant mess above, this comment can probably go away

File: sprokit/src/sprokit/pipeline/process_factory.h [hunk 12]
```diff
@@ -193,6 +194,49 @@ kwiver::vital::plugin_factory_vector_t const& get_process_list();
...
+  }
+
+  // ----------------------------------------------------------------------------
+  template <typename process_t>
+  kwiver::vital::plugin_factory_handle_t  register_process( option opt = none )
+  {
+    using kvpf = kwiver::vital::plugin_factory;
+
+    auto fact = plugin_loader().
+      add_factory( new sprokit::cpp_process_factory( typeid( process_t ).name(),
```
Comment (mwoehlke-kitware) [inline sprokit/src/sprokit/pipeline/process_factory.h]: Ugh, inconsistent indent... please start a new line after `cpp_process_factory(`. (Also, maybe use uniform initialization here instead?)

File: sprokit/src/sprokit/pipeline/process_factory.h [hunk 13]
```diff
@@ -193,6 +194,49 @@ kwiver::vital::plugin_factory_vector_t const& get_process_list();
                                                  typeid( sprokit::process ).name(), \
                                                  sprokit::create_new_process< proc_type > ) )
 
+// ============================================================================
+class process_registrar
+  : public kwiver::plugin_registrar
+{
+public:
+  enum option {
```
Comment (mwoehlke-kitware) [inline sprokit/src/sprokit/pipeline/process_factory.h]: Possibly overengineering at this point (i.e. something we can add later, at need), but this smells like a candidate for a flags type.

File: vital/plugin_loader/plugin_info.h [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef VITAL_PLUGIN_LOADER_PLUGIN_INFO_H
+#define VITAL_PLUGIN_LOADER_PLUGIN_INFO_H
+
+#define PLUGIN_INFO(NAME, DESCRIP)             \
```
Comment (mwoehlke-kitware) [inline vital/plugin_loader/plugin_info.h]: I wonder if this should instead expand to `PLUGIN_INFO_WITH_ORG(NAME, DESCRIPTION, KWIVER_DEFAULT_PLUGIN_ORGANIZATION)`? This would be more convenient if some user wants/needs to specify a different organization. 

Note: this implies making the organization also a static member, and also probably:
```c++
#ifndef KWIVER_DEFAULT_PLUGIN_ORGANIZATION
#define KWIVER_DEFAULT_PLUGIN_ORGANIZATION "Unspecified"
#endif
```
...here, rather than in `plugin_registrar.h`.

(Feel free to bikeshed on the name of the macro.)

File: tools/applet_registrar.h (hunk 2 above)
Comment (linus-sherrill) [reply tools/applet_registrar.h]: done

File: sprokit/src/sprokit/pipeline/process_factory.h (hunk 3 above)
Comment (linus-sherrill) [reply sprokit/src/sprokit/pipeline/process_factory.h]: done

File: vital/algo/algorithm_factory.h (hunk 4 above)
Comment (linus-sherrill) [reply vital/algo/algorithm_factory.h]: done

File: vital/plugin_loader/plugin_registrar.h (hunk 5 above)
Comment (linus-sherrill) [reply vital/plugin_loader/plugin_registrar.h]: done

File: vital/plugin_loader/plugin_registrar.h (hunk 6 above)
Comment (linus-sherrill) [reply vital/plugin_loader/plugin_registrar.h]: done

File: sprokit/processes/adapters/register_processes.cxx (hunk 9 above)
Comment (linus-sherrill) [reply sprokit/processes/adapters/register_processes.cxx]: process registrar is in sprokit name space

File: vital/plugin_loader/plugin_info.h (hunk 14 above)
Comment (linus-sherrill) [reply vital/plugin_loader/plugin_info.h]: Fixed the organization problem by adding it to the CMake file.

**Comment by linus-sherrill** [comment]
jenkins test this please

File: sprokit/processes/adapters/register_processes.cxx (hunk 10 above)
Comment (linus-sherrill) [reply sprokit/processes/adapters/register_processes.cxx]: done
